### PR TITLE
(test) Remove unnecessary partial mocks and improve typing

### PR DESCRIPTION
**Pull request by denniskigen**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary

This PR refactors most tests in this repository with the following goals:

**Streamlining mocks**
- **Removing unnecessary [partial mocks](https://jestjs.io/docs/mock-functions#mocking-partials)**: 
  The [mock.tsx](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) manual mock in Core aggregates stubs for most functions in Core. We should only extend the manual mock to include stubs for things that exist in Core but are not covered by the manual mock. Otherwise, we should use the existing stubs provided by the manual mock.

**Improving type safety and intellisense**
- **Replacing `const mockedSourceFn = sourceFn as jest.Mock` with `jest.mocked(sourceFn)`**: 
  The [jest.mocked](https://jestjs.io/docs/mock-function-api#jestmockedsource-options) method infers type definitions from the source function, enhancing type checking and Intellisense. This ensures our mock data closely resembles the actual source data's shape, providing more confidence in our tests.

**Automatic mock clearing**
- **Setting `clearMocks: true` in Jest config**: 
  This configuration ensures that mock calls, instances, contexts, and results are automatically cleared between tests, as detailed [here](https://jestjs.io/docs/configuration#clearmocks-boolean).

**Enhanced type inference for `useConfig` mocks**
- **Annotating `useConfig` mocks**: 
  Using types from the associated config schemas improves type inference and ensures our mocks are more accurate.

**Fixing broken tests**
- **Correcting tests**: 
  Addressing and fixing tests that were not functioning as intended.

## Other

# Testing patterns

This is a brief overview of some of the patterns extracted from this refactor that will help you write more effective, maintainable and type-safe tests:

## 1. Overriding stubs in the framework mock

When testing functionality stubbed out in the framework mock, you only need to override their behaviour using mock functions.

### Example: Mocking the `useSession` hook

```typescript
import { useSession } from '@openmrs/esm-framework';

const mockUseSession = jest.mocked(useSession);

describe('Tests using session', () => {
  beforeEach(() => {
    mockUseSession.mockReturnValue(mockSession.data);
  });
  
  // Your test assertions...
});
```

You don't need a partial mock for this scenario because the `useSession` function is already stubbed in the framework mock. This happens courtesy of this line in the root-level `jest.config.js` file:

```js
moduleNameMapper: {
  '@openmrs/esm-framework': '@openmrs/esm-framework/mock',
},
```

This line tells Jest to use the `mock.tsx` file as a mock for the `@openmrs/esm-framework` module. This mock file contains stubs for most of the functions in the framework, including `useSession`.

## 2. Extending the framework mock using partial mocks

You only extend the framework mock using partials if the functionality you're testing doesn't have a corresponding stub in the framework mock.

### How to check for existing stubs:

1. Search for the function name in the Core repository.
2. Limit the search to mock files by typing `*.mock.tsx` in the `files to include` field.

If the function you're testing doesn't have a corresponding stub in the framework mock, you can extend the framework mock using partial mocks. This should be a temporary solution until the functionality is added to the framework mock.

### Example: Mocking `useLocations` (assuming no existing stub)

```typescript
import { useLocations } from '@openmrs/esm-framework';
import { mockLocationsResponse } from '__mocks__';

const mockUseLocations = jest.mocked(useLocations);

jest.mock('@openmrs/esm-framework', () => ({
  ...jest.requireActual('@openmrs/esm-framework'),
  useLocations: jest.fn(),
}));

describe('Tests using locations', () => {
  beforeEach(() => {
    mockUseLocations.mockReturnValue(mockLocationsResponse.data);
  });

  // Your test assertions...
});
```

**Note:** Don't include stubs for existing functions in the framework mock in your partial mock.

## 3. Inferring types from mocked functions

Use `jest.mocked` instead of type assertions to preserve type information and improve type safety.

### Incorrect (loses type information):
```typescript
const mockUseLayoutType = useLayoutType as jest.Mock;
```

### Correct (preserves type information):
```typescript
const mockUseLayoutType = jest.mocked(useLayoutType);
```

This approach provides better type safety and IntelliSense support, flagging type errors when incorrect return values or mock implementations are provided.

## 4. Improving type inference for config mocks

Enhance type inference for `useConfig` mocks using `jest.mocked` with the appropriate config type.

### Example: Mocking config with type inference

```typescript
import { getDefaultsFromConfigSchema, useConfig } from '@openmrs/esm-framework';
import { type RegistrationConfig, esmPatientRegistrationSchema } from '../config-schema';

const mockUseConfig = jest.mocked(useConfig<RegistrationConfig>);

describe('Tests using config', () => {
  beforeEach(() => {
    mockUseConfig.mockReturnValue({
      ...getDefaultsFromConfigSchema(esmPatientRegistrationSchema),
      fieldConfigurations: {
        address: {
          useAddressHierarchy: {
            enabled: false,
            useQuickSearch: false,
            searchAddressByLevel: false,
          },
        },
      } as RegistrationConfig['fieldConfigurations'],
    });
  });

  // Your test assertions...
});
```

With this, if I provide an incorrect value for the useConfig mock, TypeScript will flag it as a type error:

CleanShot 2024-07-26 at 4  58 41@2x

This approach provides better type inference and IntelliSense support when working with config objects in your tests.

## Summary

1. Use existing framework mock stubs when available.
2. Only extend the framework mock for functionality without existing stubs.
3. Use `jest.mocked` to preserve the type information of mocked functions.
4. Leverage type inference for config mocks to improve type safety.
5. Keep your tests DRY (Don't Repeat Yourself) by reusing mock setups and test utilities.
6. Regularly update your local framework mock to align with the latest Core updates.

Following these patterns and best practices will help you write more robust, type-safe, and maintainable tests.

### CODE REVIEW

File: packages/esm-active-visits-app/src/visits-summary/visit-detail.test.tsx [hunk 1]
```diff
@@ -96,27 +101,22 @@ describe('VisitDetailComponent', () => {
...
-
-    render(<VisitDetailComponent visitUuid={visitUuid} patientUuid={patientUuid} />);
-
-    expect(screen.queryByRole('button', { name: 'All Encounters' })).not.toBeInTheDocument();
-  });
 });
+
+function renderVisitDetail() {
+  render(<VisitDetailComponent {...defaultProps} />);
+}
```
Comment (brandones) [inline packages/esm-active-visits-app/src/visits-summary/visit-detail.test.tsx:122]: I'm not a big fan of these kinds of functions. A one-liner like that probably doesn't need to be factored out at all. These functions are useful when the thing to be rendered involves nesting components or context or other more complex stuff.

**Review by brandones — APPROVED** [review]
I gave this a cursory look. I imagine it's fine, given that the crux of the change is to make typing more strict. But I'm sure a more thorough review couldn't hurt.

File: packages/esm-service-queues-app/src/side-menu/side-menu.test.tsx [hunk 2]
```diff
@@ -5,7 +5,6 @@ import SideMenu from './side-menu.component';
 jest.mock('@openmrs/esm-framework', () => ({
```
Comment (brandones) [inline packages/esm-service-queues-app/src/side-menu/side-menu.test.tsx:5]: I was going to point out that `LeftNavMenu` is [already mocked](https://github.com/openmrs/openmrs-esm-core/blob/4d3f1b0e33f51f805ef75ac81fe9d8e849e6e931/packages/framework/esm-framework/mock.tsx#L61), but actually I think this is maybe a useless test and we would be better off just deleting this file.

**Review by brandones — APPROVED** [review]

**Comment by denniskigen** [comment]
Good to go, @ibacher?

**Comment by ibacher** [comment]
Looks fine to me.

**Pull request merged**